src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt

```diff
@@ -0,0 +1,89 @@
+import org.springframework.security.oauth2.jwt.JwtEncoder
+import org.springframework.security.oauth2.jwt.JwtEncoderParameters
+import org.springframework.stereotype.Service
+import org.springframework.web.server.ResponseStatusException
+import pt.up.fe.ni.website.backend.config.auth.AuthConfigProperties
+import pt.up.fe.ni.website.backend.model.Account
+import java.time.Duration
+import java.time.Instant
+import java.util.stream.Collectors
+
+@Service
+class AuthService(
+    val accountService: AccountService,
+    val authConfigProperties: AuthConfigProperties,
+    val jwtEncoder: JwtEncoder,
+    val jwtDecoder: JwtDecoder,
+    private val passwordEncoder: PasswordEncoder
+) {
+    fun authenticate(email: String, password: String): Account {
+        val account = accountService.getAccountByEmail(email)
+        if (!passwordEncoder.matches(password, account.password)) {
+            throw ResponseStatusException(HttpStatus.UNPROCESSABLE_ENTITY, "invalid credentials")
+        }
+        val authentication = UsernamePasswordAuthenticationToken(email, password, getAuthorities(account))
+        SecurityContextHolder.getContext().authentication = authentication
+        return account
+    }
+
+    fun generateAccessToken(account: Account): String {
+        return generateToken(account, Duration.ofMinutes(authConfigProperties.jwtAccessExpirationMinutes))
+    }
+
+    fun generateRefreshToken(account: Account): String {
+        return generateToken(account, Duration.ofDays(authConfigProperties.jwtRefreshExpirationDays), true)
+    }
+
+    fun refreshAccessToken(refreshToken: String): String {
+        val jwt =
+            try {
+                jwtDecoder.decode(refreshToken)
+            } catch (e: Exception) {
+                throw ResponseStatusException(HttpStatus.UNAUTHORIZED, "invalid refresh token")
+            }
+        if (jwt.expiresAt?.isBefore(Instant.now()) != false) {
+            throw ResponseStatusException(HttpStatus.UNAUTHORIZED, "refresh token has expired")
+        }
+        val account = accountService.getAccountByEmail(jwt.subject)
+        return generateAccessToken(account)
+    }
+
+    fun getAuthenticatedAccount(): Account {
+        val authentication = SecurityContextHolder.getContext().authentication
+        return accountService.getAccountByEmail(authentication.name)
+    }
+
+    private fun generateToken(account: Account, expiration: Duration, isRefresh: Boolean = false): String {
+        val roles = if (isRefresh) emptyList() else getAuthorities(account)
+        val scope = roles
+            .stream()
+            .map(GrantedAuthority::getAuthority)
+            .collect(Collectors.joining(" "))
+        val currentInstant = Instant.now()
+        val claims = JwtClaimsSet
+            .builder()
+            .issuer("self")
+            .issuedAt(currentInstant)
+            .expiresAt(currentInstant.plus(expiration))
+            .subject(account.email)
+            .claim("scope", scope)
```

Comment:
If I understood well, the claim is composed of a scope which shows the user's roles?

Reply:
Yes, in this implementation yes. There is some discussion against it: https://stackoverflow.com/questions/47224931/is-setting-roles-in-jwt-a-best-practice. However, in our use case, to reduce database fetches and code complexity, I think it is a worthy, small security tradeoff.

Reply:
I see, I don't think it should be a big concern for us